### ISSUE TITLE
make error

### ISSUE THREAD

**Issue by arjunmenon** [rpav/c2ffi#56]
Hey
I cant seem to build c2ffi with clang 5.0
```java
>:~/Desktop/ruby/c2ffi/build$ cmake ..
-- Found LLVM 5.0.2
-- LLVM installed in /usr/lib/llvm-5.0
-- Using LLVMConfig.cmake in: /usr/lib/llvm-5.0/cmake
-- Using llvm libs: LLVMCore;LLVMSupport;LLVMMCParser;LLVMOption;LLVMBitReader;LLVMProfileData
-- CMAKE_CXX_FLAGS:  -std=c++11 -fno-exceptions -fno-rtti
-- CMAKE_EXE_LINKER_FLAGS: 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/arjun/Desktop/ruby/c2ffi/build

>:~/Desktop/ruby/c2ffi/build$ make
Scanning dependencies of target c2ffi
[  7%] Building CXX object CMakeFiles/c2ffi.dir/src/Expr.cpp.o
/home/arjun/Desktop/ruby/c2ffi/src/Expr.cpp:22:29: fatal error: clang/Lex/Token.h: No such file or directory
compilation terminated.
CMakeFiles/c2ffi.dir/build.make:62: recipe for target 'CMakeFiles/c2ffi.dir/src/Expr.cpp.o' failed
make[2]: *** [CMakeFiles/c2ffi.dir/src/Expr.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/c2ffi.dir/all' failed
make[1]: *** [CMakeFiles/c2ffi.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

I installed clang5.0 using this guide - https://askubuntu.com/questions/905205/installing-clang-5-0-and-using-c17

```java
>$ dpkg -l llvm*
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                                 Version                 Architecture            Description
+++-====================================-=======================-=======================-==============================================================================
un  llvm                                 <none>                  <none>                  (no description available)
un  llvm-3.8-dev                         <none>                  <none>                  (no description available)
ii  llvm-5.0                             1:5.0.2~svn328729-1~exp amd64                   Modular compiler and toolchain technologies
ii  llvm-5.0-dev                         1:5.0.2~svn328729-1~exp amd64                   Modular compiler and toolchain technologies, libraries and headers
un  llvm-5.0-doc                         <none>                  <none>                  (no description available)
ii  llvm-5.0-runtime                     1:5.0.2~svn328729-1~exp amd64                   Modular compiler and toolchain technologies, IR interpreter
```

I am on elementary os 0.4/ubuntu 16.04 build

**Comment by rpav** [comment]
It's possible/probable the package puts headers in a strange location, I'd try something like:

```console
$ dpkg -L llvm-5.0-dev | grep Lex/Token.h
```

See if/where it puts it, possibly adjust CXXFLAGS accordingly.

**Comment by arjunmenon** [comment]
I installed `libclang-5.0-dev` from the repository. It started working.

Though, after generating json files, I could not use the c2ffi-ruby gem to create FFI bindings. I started getting segmentation faults and even the generated ffi file had many errors.

**Comment by rpav** [comment]
The c2ffi-ruby stuff was an early, simple implementation to proof that this worked in another language. It didn't spark a lot of interest and I don't really do Ruby anymore, so it's unmaintained and probably far behind in many respects (both on the c2ffi and ruby sides).

That said, it should probably not be hard to update or write something similar.

**Comment by arjunmenon** [comment]
Alright, how can we utilize the Json files to create bindings manually or is it possible?

**Comment by rpav** [comment]
Well... you can use them however you'd like?  If you just want a few simple functions, I'm not sure even this is necessary.  But probably you want to write a binding generator that just reads this file and makes the appropriate API calls.

If so I'd highly recommend spitting out an intermediate, formatted, VCS-friendly version of the data, and not storing `c2ffi` output directly.

**Comment by arjunmenon** [comment]
> If so I'd highly recommend spitting out an intermediate, formatted, VCS-friendly version of the data, and not storing c2ffi output directly.

That had a _whoosh_ effect on top of my head.

Do let me know if a simple solution like your lib exists to make the JSON output workable.

**Comment by rpav** [comment]
The JSON output is usable as-is; I'd start there.  [cl-autowrap](https://github.com/rpav/cl-autowrap) uses this.  A lot of the features of the JSON were designed around using it this way.

With ruby-ffi for instance, it looks like you'd want to read through the JSON and generate a bunch of `layout` and `attach_function` as appropriate.  There are a few things that make this slightly nontrivial (anonymous records tend to stand out), but for the most part it's straightforward.  This will give you FFI bindings.

However, `c2ffi` output varies textually, based on what system you run it, how the local headers are set up, etc.  It includes local paths for each line, so generating errors can be informative.  It also includes *everything*.  This tends to not be terribly version-control friendly, and the files tend to be quite a bit larger than what's strictly necessary.

If you spit out ruby source files and discard the `c2ffi` output, you can mostly ignore that... though you may want to generate "stable" source.

If like autowrap you keep the output around, you may eventually want to generate your own, more stable, minimized files.

**tl;dr**: Start simple with the JSON as-is.  Later, maybe optimize.
